flink-streaming-java/src/test/java/org/apache/flink/streaming/util/HDFSCopyUtilitiesTest.java

```diff
@@ -70,6 +70,85 @@ public void testCopyFromLocal() throws Exception {
 		}
 	}
 
+
+	/**
+	 * This test verifies that nested directories are properly copied.
+	 */
+	@Test
+	public void testCopyFromLocalRecursive() throws Exception {
+
+		File rootDir = tempFolder.newFolder();
+		File nestedDir = new File(rootDir,"nested");
+		nestedDir.mkdir();
+
+		Map<String,File>  copyFiles = new HashMap<String,File>();
+
+		copyFiles.put("1",new File(rootDir, "1"));
+		copyFiles.put("2",new File(rootDir, "2"));
+		copyFiles.put("3",new File(nestedDir, "3"));
+
+		for (File file : copyFiles.values()) {
+			try (DataOutputStream out = new DataOutputStream(new FileOutputStream(file))) {
+				out.writeUTF("Hello there, " + file.getName());
+			}
+		}
+		//add root and nested dirs to expected output
+		copyFiles.put(rootDir.getName(),rootDir);
+		copyFiles.put("nested",nestedDir);
+
+		assertEquals(5,copyFiles.size());
+
+		//Test for copy to unspecified target directory
+		File copyDirU = tempFolder.newFolder();
+		HDFSCopyFromLocal.copyFromLocal(
+				rootDir,
+				new Path("file://" + copyDirU.getAbsolutePath()).toUri());
+
+		//Test for copy to specified target directory
+		File copyDirQ = tempFolder.newFolder();
+		HDFSCopyFromLocal.copyFromLocal(
+				rootDir,
+				new Path("file://" + copyDirQ.getAbsolutePath() + "/" + rootDir.getName()).toUri());
+
+		FilenameFilter noCrc = new FilenameFilter() {
```

Comment:
Checking crc or other FileSystem-maintained shadow files would complicate the test. We just need to verify that the file structure we intended to copy is in fact copied exactly once.